cranelift/codegen/src/isa/s390x/inst/mod.rs

```diff
@@ -1470,55 +1552,62 @@ impl Inst {
                 let rd = pretty_print_reg_mod(rd, ri, allocs);
                 format!("{} {}, {}", op, rd, imm.bits)
             }
-            &Inst::SMulWide { rn, rm } => {
+            &Inst::SMulWide { rd, rn, rm } => {
                 let op = "mgrk";
                 let rn = pretty_print_reg(rn, allocs);
                 let rm = pretty_print_reg(rm, allocs);
-                let rd = pretty_print_reg(gpr(0), allocs);
-                let _r1 = allocs.next(gpr(1));
-                format!("{} {}, {}, {}", op, rd, rn, rm)
+                let rd1 = pretty_print_reg(w_regpair_hi(rd).to_reg(), allocs);
+                let rd2 = pretty_print_reg(w_regpair_lo(rd).to_reg(), allocs);
+                format!("{} {}, {}, {}, {}", op, rd1, rd2, rn, rm)
```

Comment:
So these changes I really do not like, for the same reason originally discussed here: https://github.com/bytecodealliance/wasmtime/pull/4856#discussion_r962932362

I really think the final assembly output needs to remain *valid* s390x assembler text, that could be run through the system assembler.  We fixed that problem originally by using a `pretty_print_reg_mod` helper that prints both virtual regs before regalloc, and just the single physreg after regalloc.

Similarly, I think we should define `pretty_print_regpair` and/or `pretty_print_regpair_mod` helpers that print all participating virtual regs before regalloc, and just the single physreg  that ends up in the assembler text after regalloc.

That should also remove many of the changes to the filetests.

Reply:
The other backends had drifted from being valid assembly, and I had assumed that the same was true of the s390x backend; I'll remove the changes to the pretty-printer :+1:

Reply:
Now that I've reverted everything back to how it was before, I'll have a look at adding the helper that you mentioned to print  the additional regs when they don't have a physical allocation.

Reply:
I made the changes to introduce `pretty_print_regpair`, `pretty_print_regpair_mod`, and `pretty_print_regpair_mod_lo`. This definitely cleaned up the pretty-printing code quite a bit, so thank you for the suggestion!

I also opted to make `RegPair` and `WritableRegPair` structs instead of enums, and marking them as `primitive` types in ISLE. This required moving some functions to rust like `regpair`, `regpair_lo`, etc, but made working with their values in rust much nicer.

Reply:
Yes, this is looking really nice now, thanks!